pipeline/utils/github_utils.py

```diff
@@ -0,0 +1,137 @@
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utility functions related to GitHub interaction. Uses the GitHub API with
+reference https://developer.github.com/v3/"""
+
+# How to update requirements so that these imports are possible?
+import requests
+import json
+import os
+import base64
+
+
+_API_URL = 'https://api.github.com'
+
+# file modes for different git file types
+_TREE_EL_MODES = {'file': u'100644',
+                  'executable': u'100755',
+                  'tree': u'040000',
+                  'symlink': u'120000'}
+
+
+def _is_file(el):
+    return os.path.isfile(el) and not os.access(el, os.X_OK)
+
+
+def _is_executable(el):
+    return os.path.isfile(el) and os.access(el, os.X_OK)
+
+
+def _get_entries(entry_type_func, a_dir):
+    """List entries in a_dir that are of type specified by entry_type_func"""
+    return filter(lambda x: entry_type_func(os.path.join(a_dir, x)),
+                  os.listdir(a_dir))
+
+
+def _get_request_content(req_func, url, req_data=None, username=None,
+                         password=None):
+    """Make an HTTP request with given data and (username, password) tuple.
+    Throw an exception if the request fails"""
+    if username and password:
+        attempt = req_func(url, data=req_data, auth=(username, password))
+    else:
+        attempt = req_func(url, data=req_data)
+    if attempt.ok:
+        return json.loads(attempt.content)
+    else:
+        raise Exception('Failed request at {0}.'
+                        'Reason: {1}'.format(url, attempt.content))
+
+
+def _build_git_tree_blob_item(file_or_exec, blob_mode, curr_dir, username,
+                              password, owner, repo):
+    """Create a blob item that can be used to create a git tree"""
+    with open(os.path.join(curr_dir, file_or_exec), 'r') as f:
+        path_content = f.read()
+    path_content = base64.b64encode(path_content)
+    blob = _get_request_content(requests.post,
+                                '{0}/repos/{1}/{2}'
+                                '/git/blobs'.format(_API_URL, owner, repo),
+                                json.dumps({'content': path_content,
+                                            'encoding': 'base64'}),
+                                username, password)
+    return {'path': file_or_exec,
+            'mode': _TREE_EL_MODES[blob_mode],
+            'type': 'blob',
+            'sha': blob['sha']}
+
+
+# TODO: Have this function also upload symlinks to the repository
+# according to GitHub protocol
+def _build_git_tree(curr_dir, username, password, owner, repo):
+    """Create a git tree from curr_dir for the given GitHub repo"""
```

Comment:
We discussed this offline-- we decided that because its subtrees have to be created before the tree object is created, os.walk is not appropriate here.